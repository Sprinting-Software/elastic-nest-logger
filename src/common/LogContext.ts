export interface LogContext {
  tenantId: string | number
  userId?: string | number
}
