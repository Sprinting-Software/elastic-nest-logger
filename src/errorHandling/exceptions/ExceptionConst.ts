let truncationLimit = 8445 // Max message length for UDP
export const ExceptionConst = {
  getTruncationLimitForExceptions: function () {
    return truncationLimit
  },
  overrideTruncationLimitForExceptions: function (limit: number) {
    truncationLimit = limit
  },
}
