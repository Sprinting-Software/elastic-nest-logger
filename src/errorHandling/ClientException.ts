import { AppException } from "./AppException"
import { HttpStatus } from "@nestjs/common"

/**
 * An exception using HTTP status code 500 InternalServerError.
 */
export class ClientException extends AppException {
  constructor(
    public readonly errorName: string,
    public readonly description?: string,
    public readonly contextData?: Record<string, any>,
    public readonly innerError?: Error
  ) {
    super(HttpStatus.BAD_REQUEST, errorName, description, contextData, innerError)
    Object.setPrototypeOf(this, ClientException.prototype)
  }
}
