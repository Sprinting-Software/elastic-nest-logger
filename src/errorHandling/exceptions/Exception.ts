import util from "util"
import { HttpStatus } from "@nestjs/common"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

export interface AppExceptionResponse {
  statusCode: number
  errorName: string
  errorTraceId: string
  message?: string | object
  contextData?: Record<string, any>
  innerError?: Error
}
export interface AppExceptionResponseV2 {
  httpStatus: number
  errorName: string
  errorTraceId: string
  message?: string | object
  contextData?: Record<string, any>
  innerError?: Error
}

const INSPECT_DEPTH = 7
const INSPECT_SHOW_HIDDEN = false
const MSG_LENGTH = 8445 // Max message length for UDP

export class Exception extends Error {
  public readonly errorTraceId: string
  constructor(
    public readonly httpStatus: number = HttpStatus.BAD_REQUEST,
    public errorName: string,
    public description?: string,
    public contextData: Record<string, any> = {},
    public innerError?: Error
  ) {
    super(errorName ?? HttpStatus[httpStatus])
    this.errorName = errorName ?? HttpStatus[httpStatus]
    this.errorTraceId = Exception.generateErrorTraceId()
    this.refreshMessageField()
    Object.setPrototypeOf(this, Exception.prototype)
  }

  public getStatus(): number {
    return this.httpStatus
  }
  override toString(): string {
    try {
      let msg = `${this.constructor.name} ${this.concatAllRelevantInfo()}`
      if (this.stack) msg += `\n ${this.generateStacktrace()}`
      if (this.innerError) {
        msg += "\nINNER_ERROR:\n"
        if (this.innerError instanceof Exception) {
          msg += this.innerError.toString()
        } else {
          const e0 = util.inspect(this.innerError)
          msg += e0
        }
      }
      if (msg.length > MSG_LENGTH) {
        return `${msg.substring(0, MSG_LENGTH)}...TRUNCATED`
      }
      return msg
    } catch (e) {
      return "ERROR_IN_TO_STRING"
    }
  }
  public generateStacktrace(): string {
    try {
      return this.stack?.split("\n").slice(1).join("\n")
    } catch (e) {
      return "STACKTRACE_GENERATION_FAILED"
    }
  }

  /**
   * Used to generate a string that can be used as a message field in ELK and in console output
   * @private
   */
  private concatAllRelevantInfo() {
    const { errorName, httpStatus, description, contextData } = this
    let msg = `${errorName} (HTTP_STATUS ${httpStatus}) (${this.errorTraceId}) `
    if (description) msg += `ERROR_DESCRIPTION - ${description}`
    if (contextData) msg += ` - CONTEXT_DATA: ${util.inspect(contextData, INSPECT_SHOW_HIDDEN, INSPECT_DEPTH)}`
    return msg
  }

  /**
   * We need to refresh the message field because we have to support the fluent API
   * where not all fields are necessarily passed in the constructor.
   * @private
   */
  private refreshMessageField() {
    this.message = this.concatAllRelevantInfo()
  }

  /**
   * Use this method to add an inner error to the current AppException object.
   * An inner error is an error that caused the current error to occur.
   *
   * @param {Error} innerError - The inner error to add to the AppException object.
   * @returns {Exception} - The modified AppException object with the new inner error.
   */
  setInnerError(innerError: Error): Exception {
    this.innerError = innerError
    this.refreshMessageField()
    return this
  }

  /*
   * Adds additional context data to the response object. Existing context data will be merged with the new data.
   * @param {Record<string, any>} contextData - The context data to be included with the error.
   * @returns {AppException} - The modified AppException object with the enriched context data
   */
  setContextData(contextData: Record<string, any>) {
    this.contextData = { ...this.contextData, ...contextData }
    this.refreshMessageField()
    return this
  }

  /**
   * Returns the response object that will be sent to the client.
   * Please do not change the method name as it matches with the NestJS built-in error interface.
   */
  getResponse(): AppExceptionResponse | AppExceptionResponseV2 {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        message: this.description,
        contextData: this.contextData,
        errorTraceId: this.errorTraceId,
      }
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        message: this.description,
        contextData: this.contextData,
        errorTraceId: this.errorTraceId,
      }
    }
  }

  /**
   * Generates a random string that can be used as an error trace id.
   * Should only contain upper case letters that cannot be confused with numbers.
   * @private
   */
  private static generateErrorTraceId() {
    const result = new Array(6)
    function getRandom(cs: string = CHARS) {
      return cs.charAt(Math.floor(Math.random() * cs.length))
    }
    for (let i = 0; i < 6; i++) {
      result[i] = getRandom(CHARS)
    }
    return `ERR${result.join("")}`
  }
}
const CHARS = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"
