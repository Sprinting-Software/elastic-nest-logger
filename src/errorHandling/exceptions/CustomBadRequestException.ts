import { BadRequestException } from "@nestjs/common"
import { Exception, AppExceptionResponse, AppExceptionResponseV2 } from "./Exception"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

export class CustomBadRequestException extends Exception {
  public readonly errors: Record<string, string[]> = {}

  constructor(private readonly badRequestException: BadRequestException) {
    super(400, badRequestException.name, badRequestException.message)

    Object.setPrototypeOf(this, CustomBadRequestException.prototype)
    this.errors = badRequestException.getResponse()["message"]
  }

  override toString(): string {
    const { errorName, httpStatus } = this
    let message = `${errorName} (HTTP_STATUS ${httpStatus})`
    message += ` ERROR_NAME - ${this.errorName}`
    message += ` ERROR_MESSAGE - ${this.description}`

    if (this.errors) message += ` VALIDATION_ERRORS - ${JSON.stringify(this.errors)}`

    return message
  }

  override getResponse() {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        message: this.errors,
      } as AppExceptionResponseV2
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        message: this.errors,
      } as AppExceptionResponse
    }
  }
}
