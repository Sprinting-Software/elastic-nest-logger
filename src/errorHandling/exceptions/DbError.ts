import { Exception } from "./Exception"

export class DbError extends Exception {
  public operationName: string
  public args: any[]

  constructor(operationName: string, args: any[] = []) {
    super(500, `Database operation ${operationName} failed`)
    this.operationName = operationName
    this.args = args
    Object.setPrototypeOf(this, DbError.prototype)
  }
}
