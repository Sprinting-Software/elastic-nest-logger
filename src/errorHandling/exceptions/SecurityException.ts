import { AppExceptionResponse, AppExceptionResponseV2, Exception } from "./Exception"
import { HttpStatus } from "@nestjs/common"
import { LibraryVersioning } from "../../libVersioning/LibraryVersioning"

/**
 * An exception using HTTP status code 403 FORBIDDEN.
 * Exceptions of this type are treated specially in that the context data, stacktrace and description
 * are never returned to the client, not even in non-production.
 * This is to ensure a minimum amount of information leakage.
 */
export class SecurityException extends Exception {
  constructor(
    public readonly description?: string,
    public readonly contextData: Record<string, any> = {},
    public readonly innerError?: Error
  ) {
    super(HttpStatus.FORBIDDEN, "SecurityException", description, contextData, innerError)
    Object.setPrototypeOf(this, SecurityException.prototype)
  }

  /**
   * Make sure details about security errors are never returned to the client.
   */
  override getResponse(): AppExceptionResponse | AppExceptionResponseV2 {
    if (LibraryVersioning.v2IsActive()) {
      return {
        httpStatus: this.httpStatus,
        errorName: this.errorName,
        errorTraceId: this.errorTraceId,
      }
    } else {
      return {
        statusCode: this.httpStatus,
        errorName: this.errorName,
        errorTraceId: this.errorTraceId,
      }
    }
  }
}
