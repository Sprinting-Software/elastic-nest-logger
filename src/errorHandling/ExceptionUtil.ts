import { Exception } from "./exceptions/Exception"
import { CustomBadRequestException } from "./exceptions/CustomBadRequestException"
import { BadRequestException } from "@nestjs/common"
import util from "util"
import { IExceptionJson } from "./IExceptionJson"
import { ServerException } from "./exceptions/ServerException"

export class ExceptionUtil {
  /**
   * Parses an error and returns an AppException.
   * @param error
   */
  public static parse(error: Error): Exception {
    if (error instanceof Exception) {
      return error
    }

    if (error.name === "BadRequestException") {
      return new CustomBadRequestException(<BadRequestException>error)
    }

    if ("getStatus" in error && typeof error.getStatus === "function") {
      return new Exception(error.getStatus(), error.name, error.message).setInnerError(error)
    }

    return new ServerException(error.name, error.message, undefined, error)
  }

  /*
  Sometimes it is useful to convert an Error to a plain json structure, for instance when you want to do snapshot-testing
  in which case it is not working for real error objects.
   */
  public static toPlainJsonForSpec(error: unknown) {
    return this.toPlainJson(error, true)
  }

  private static toPlainJson(error: unknown, redactTraceFields = false): IExceptionJson {
    if (!error) return { errorName: "UndefinedError", description: `Found this error: ${util.inspect(error)}` }

    if (error instanceof Exception) {
      const exc = {
        errorName: error.errorName,
        description: error.description,
        contextData: error.contextData,
        httpStatus: error.httpStatus,
        errorTraceId: redactTraceFields ? "REDACTED" : error.errorTraceId,
      } as IExceptionJson
      if (error.innerError) exc.innerError = this.toPlainJson(error.innerError, redactTraceFields)
      return exc as IExceptionJson
    }
    if (error instanceof Error) return { errorName: error.name, description: error.message }
    return { errorName: "UnknownError", description: `Found this error: ${util.inspect(error)}` }
  }
}
