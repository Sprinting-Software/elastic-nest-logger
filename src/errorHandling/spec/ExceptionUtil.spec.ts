import { Exception } from "../exceptions/Exception"
import { HttpStatus } from "@nestjs/common"
import { ClientException } from "../exceptions/ClientException"
import { ExceptionUtil } from "../ExceptionUtil"
import { ServerException } from "../exceptions/ServerException"
import { SecurityException } from "../exceptions/SecurityException"
import { AssertionException } from "../exceptions/AssertionException"

describe("ExceptionUtil", () => {
  it("should parse ClientException", () => {
    const e = new ClientException("ERROR_NAME", "ERROR_DESCRIPTION", { key: "value" })
    const e2 = ExceptionUtil.parse(e)
    expect(e2).toBe(e)
    expect(e2).toEqual(e)
    expect(e2 === e).toBeTruthy()
  })
  it("should parse AppException", () => {
    const e = getComplexError()
    const e2 = ExceptionUtil.parse(e)
    expect(e2).toBe(e)
    expect(e2).toEqual(e)
    expect(e2 === e).toBeTruthy()
  })
  it("should parse ClientException", () => {
    const e = new ServerException("ERROR_NAME", "ERROR_DESCRIPTION", { key: "value" })
    const e2 = ExceptionUtil.parse(e)
    expect(e2).toBe(e)
    expect(e2).toEqual(e)
    expect(e2 === e).toBeTruthy()
  })

  it("should create plain objects from complex error 1", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getComplexError())).toMatchSnapshot()
  })
  it("should create plain objects from complex error 2", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getComplexError2())).toMatchSnapshot()
  })
  it("should create plain objects from complex error 3", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getComplexError3())).toMatchSnapshot()
  })
  it("should create plain objects from normal error", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getNormalError())).toMatchSnapshot()
  })

  it("should create plain objects from ClientException", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getClientException())).toMatchSnapshot()
  })

  it("should create plain objects from AssertionException", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getAssertionException())).toMatchSnapshot()
  })

  it("should create plain objects from SecurityException", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getSecurityException())).toMatchSnapshot()
  })

  it("should create plain objects from ServerException", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getServerException())).toMatchSnapshot()
  })

  it("should create plain objects from irregularError", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getIrregularError())).toMatchSnapshot()
  })
  it("should create plain objects from irregularError2", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(getIrregularError2())).toMatchSnapshot()
  })

  it("should create plain objects from null error", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(null)).toMatchSnapshot()
  })
  it("should create plain objects from undefined error", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(undefined)).toMatchSnapshot()
  })
  it("should create plain objects from false error", () => {
    expect(ExceptionUtil.toPlainJsonForSpec(false)).toMatchSnapshot()
  })
})

function getComplexError() {
  return new Exception(HttpStatus.INTERNAL_SERVER_ERROR, "ERROR_NAME", "ERROR_DESCRIPTION", { key: "value" })
}
function getComplexError2() {
  return new Exception(
    HttpStatus.INTERNAL_SERVER_ERROR,
    "ERROR_NAME",
    "ERROR_DESCRIPTION",
    { key: "value" },
    new Error("Some Inner Error")
  )
}
function getComplexError3() {
  return new Exception(
    HttpStatus.INTERNAL_SERVER_ERROR,
    "ERROR_NAME",
    "ERROR_DESCRIPTION",
    { key: "value" },
    getComplexError2()
  )
}

function getSecurityException() {
  return new SecurityException("ERROR_DESCRIPTION", { key: "value" }, new Error("Some Inner Error"))
}
function getAssertionException() {
  return new AssertionException("ERROR_DESCRIPTION", { key: "value" }, new Error("Some Inner Error"))
}
function getClientException() {
  return new ClientException("SOME_ERROR_NAME", "ERROR_DESCRIPTION", { key: "value" }, new Error("Some Inner Error"))
}

function getServerException() {
  return new ServerException("SOME_ERROR_NAME", "ERROR_DESCRIPTION", { key: "value" }, new Error("Some Inner Error"))
}

function getNormalError() {
  return new Error("Some normal error")
}

function getIrregularError() {
  return "Some error as string"
}

function getIrregularError2() {
  return { errorName: "Some error as object" }
}
