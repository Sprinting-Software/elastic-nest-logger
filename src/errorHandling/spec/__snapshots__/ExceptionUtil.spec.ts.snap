// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExceptionUtil should create plain objects from AssertionException 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "AssertionException",
  "errorTraceId": "REDACTED",
  "httpStatus": 502,
  "innerError": {
    "description": "Some Inner Error",
    "errorName": "Error",
  },
}
`;

exports[`ExceptionUtil should create plain objects from ClientException 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "SOME_ERROR_NAME",
  "errorTraceId": "REDACTED",
  "httpStatus": 400,
  "innerError": {
    "description": "Some Inner Error",
    "errorName": "Error",
  },
}
`;

exports[`ExceptionUtil should create plain objects from SecurityException 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "SecurityException",
  "errorTraceId": "REDACTED",
  "httpStatus": 403,
  "innerError": {
    "description": "Some Inner Error",
    "errorName": "Error",
  },
}
`;

exports[`ExceptionUtil should create plain objects from ServerException 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "SOME_ERROR_NAME",
  "errorTraceId": "REDACTED",
  "httpStatus": 500,
  "innerError": {
    "description": "Some Inner Error",
    "errorName": "Error",
  },
}
`;

exports[`ExceptionUtil should create plain objects from complex error 1 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "ERROR_NAME",
  "errorTraceId": "REDACTED",
  "httpStatus": 500,
}
`;

exports[`ExceptionUtil should create plain objects from complex error 2 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "ERROR_NAME",
  "errorTraceId": "REDACTED",
  "httpStatus": 500,
  "innerError": {
    "description": "Some Inner Error",
    "errorName": "Error",
  },
}
`;

exports[`ExceptionUtil should create plain objects from complex error 3 1`] = `
{
  "contextData": {
    "key": "value",
  },
  "description": "ERROR_DESCRIPTION",
  "errorName": "ERROR_NAME",
  "errorTraceId": "REDACTED",
  "httpStatus": 500,
  "innerError": {
    "contextData": {
      "key": "value",
    },
    "description": "ERROR_DESCRIPTION",
    "errorName": "ERROR_NAME",
    "errorTraceId": "REDACTED",
    "httpStatus": 500,
    "innerError": {
      "description": "Some Inner Error",
      "errorName": "Error",
    },
  },
}
`;

exports[`ExceptionUtil should create plain objects from false error 1`] = `
{
  "description": "Found this error: false",
  "errorName": "UndefinedError",
}
`;

exports[`ExceptionUtil should create plain objects from irregularError 1`] = `
{
  "description": "Found this error: 'Some error as string'",
  "errorName": "UnknownError",
}
`;

exports[`ExceptionUtil should create plain objects from irregularError2 1`] = `
{
  "description": "Found this error: { errorName: 'Some error as object' }",
  "errorName": "UnknownError",
}
`;

exports[`ExceptionUtil should create plain objects from normal error 1`] = `
{
  "description": "Some normal error",
  "errorName": "Error",
}
`;

exports[`ExceptionUtil should create plain objects from null error 1`] = `
{
  "description": "Found this error: null",
  "errorName": "UndefinedError",
}
`;

exports[`ExceptionUtil should create plain objects from undefined error 1`] = `
{
  "description": "Found this error: undefined",
  "errorName": "UndefinedError",
}
`;
